style_guides/js_style_guide.md

```diff
@@ -153,188 +185,288 @@ Variables, properties and function names should use `lowerCamelCase`.  They
+class bank_account {}
+class bankAccount {}
 ```
 
-*Wrong:*
+## Magic numbers/strings
+
+These are numbers (or other values) simply used in line in your code. *Do not
+use these*, give them a variable name so they can be understood and changed
+easily.
 
 ```js
-function bank_Account() {
-}
-```
+// good
+const minWidth = 300;
 
-## Use UPPERCASE for Constants
+if (width < minWidth) {
+  ...
+}
 
-Constants should be declared as regular variables or static class properties,
-using all uppercase letters.
+// bad
+if (width < 300) {
+  ...
+}
+```
 
-Node.js / V8 actually supports mozilla's [const][const] extension, but
-unfortunately that cannot be applied to class members, nor is it part of any
-ECMA standard.
+## Object properties and functions
 
-*Right:*
+Use object method shorthand syntax for functions on objects:
 
 ```js
-var SECOND = 1 * 1000;
+// good
+const foo = {
+  bar() {
+    ...
+  }
+};
 
-function File() {
-}
-File.FULL_PERMISSIONS = 0777;
+// bad
+const foo = {
+  bar: function () {
+    ...
+  }
+};
 ```
 
-*Wrong:*
+Use property value shorthand syntax for properties that share a name with a
+variable. And put them at the beginning:
 
 ```js
-const SECOND = 1 * 1000;
+const bar = true;
 
-function File() {
-}
-File.fullPermissions = 0777;
-```
+// good
+const foo = {
+  bar
+};
 
-[const]: https://developer.mozilla.org/en/JavaScript/Reference/Statements/const
+// bad
+const foo = {
+  bar: bar
+};
 
-## Magic numbers
+// also bad
+const foo = {
+  baz: false,
+  bar
+};
+```
 
-These are numbers (or other values) simply used in line in your code. **Do not use these**, give them a variable name so they can be understood and changed easily.
+## Modules
 
-*Right:*
+Module dependencies should be written using native ES2015 syntax wherever
+possible (which is almost everywhere):
 
 ```js
-var minWidth = 300;
+// good
+import { mapValues } from 'lodash';
+export default mapValues;
 
-if (width < minWidth) {
+// bad
+const _ = require('lodash');
+module.exports = _.mapValues;
+
+// worse
+define(['lodash'], function (_) {
   ...
-}
+});
 ```
 
-*Wrong:*
+In those extremely rare cases where you're writing server-side JavaScript in a
+file that does not pass run through webpack, then use CommonJS modules.
+
+In those even rarer cases where you're writing client-side code that does not
+run through webpack, then do not use a module loader at all.
+
+## Import only top-level modules
+
+The files inside a module are implementation details of that module. They
+should never be imported directly. Instead, you must only import the top-level
+API that's exported by the module itself.
+
+On the other hand, a module should be able to import parent and sibling
+modules.
 
 ```js
-if (width < 300) {
-  ...
-}
+// good
+import foo from 'foo';
+import child from './child';
+import parent from '../';
+import ancestor from '../../../';
+import sibling from '../foo';
+
+// bad
+import inFoo from 'foo/child';
+import inSibling from '../foo/child';
```

Comment:
I don't understand the reasoning here. Perhaps this could use more elaborate examples and explanations of the problems encountered when you import things this way?

Also, is it possible that this is more of a "code smell" that hints at a flawed module design?